### PR TITLE
feat: Seq scanner scans data by time range

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
https://github.com/GreptimeTeam/greptimedb/pull/4757

## What's changed and what's your intention?

Seq scanner scans data according to the order of partition ranges.

Moves `scan_file_ranges` and `scan_mem_ranges` to `scan_util`. Refactors them so that `SeqScan` and `UnorderedScan` can reuse them.

To reuse codes, this PR adds a new metric struct `PartitionMetrics`, and shares it between streams in the same partition. This struct also prints the debug log in `drop()` so we can still get the log when the stream is dropped before exhausted.

It removes all unused codes.

There are still some remaining works:
- Remove the scan parallelism. But we need to support file-level parallelism.
- Support splitting multiple row groups in `SeqScan`.
- Support field pruning in last_non_null mode if the time range only has one file.
- More tests for `RangeMeta` and `StreamContext`

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/mito2/src/read/seq_scan.rs [hunk 1]
```diff
@@ -525,292 +290,66 @@ impl RegionScanner for SeqScan {
...
-        for ranges in part.file_ranges[0].chunks(ranges_per_part) {
-            let new_part = ScanPart {
-                memtable_ranges: Vec::new(),
-                file_ranges: smallvec![ranges.to_vec()],
-                time_range: part.time_range,
+            let read_type = if compaction {
+                "compaction"
+            } else {
+                "seq_scan_files"
             };
```
Comment (waynexia) [inline src/mito2/src/read/seq_scan.rs:331]: why not use `get_scanner_type` here

**Review by waynexia — APPROVED** [review]

File: src/mito2/src/read/seq_scan.rs (hunk 1 above)
Comment (evenyag) [reply src/mito2/src/read/seq_scan.rs:331]: It doesn't return the expected label value for this metric.

**Comment by evenyag** [comment]
https://github.com/GreptimeTeam/greptimedb/pull/4809/commits/74b4088b325cd2ff1dffacd693e58512b9d47a8a changed the `PartitionRange::end` to exclusive. @waynexia @discord9

File: src/mito2/src/read/range.rs [hunk 2]
```diff
@@ -264,3 +287,77 @@ fn maybe_split_ranges_for_seq_scan(ranges: Vec<RangeMeta>) -> Vec<RangeMeta> {
...
+                (1, 1000, 2000),
+                (2, 6000, 7000),
+                (3, 0, 1500),
+                (4, 1500, 3000),
+            ],
+            &[(vec![3, 1, 4], 0, 3000), (vec![2], 6000, 7000)],
+        );
+
+        // 1, 2, 3 => [3], [1], [2],
+        run_group_ranges_test(
```
Comment (discord9) [inline src/mito2/src/read/range.rs:354]: Maybe add a `(1, 3000, 4000), (2, 4000, 5000)` case which will concat to `[1,2]`?

**Review by discord9 — APPROVED** [review]
others LGTM

**Pull request merged**